general/development/tools/behat.md

```diff
@@ -0,0 +1,721 @@
+
+- Behat requirement not satisfied: http://127.0.0.1/m/stable_master is not available, ensure you specified correct url and that the server is set up and started.
+- `init.php` or `util.php` complain that `"Unknown error 1 This is not a behat test site!"`.  Delete the behat `wwwroot` (look in `config.php` for `$CFG->behat_dataroot`) and drop the behat DB tables (look in `config.php` for `$CFG->behat_prefix`).  Then try again.
+- Behat is configured but not enabled on this test site.
+In order to fix those errors please check that: the `behat_dataroot` has correct write permissions and that the `$CFG->behat*` variables are placed before the `lib/setup.php` include:
+
+```php
+ require_once(__DIR__ . '/lib/setup.php');
+```
+
+### Selenium server is not running
+
+#### Chrome specific
+
+If you are using chrome, you need to ensure that the driver matches the version of the installed chrome browser – which may change on OS updates/upgrades. Moodle or Selenium will not give the appropriate message – see [MDL-67659](https://tracker.moodle.org/browse/MDL-67659/). One solution is the one suggested in the issue and use Andrew Nicols' [Chromedriver Wrapper](https://github.com/andrewnicols/chromedriver-wrapper/) which will ensure you have the appropriate driver before running the tests.
+
+## External links
+
+- Vagrant profile with Moodle and Behat preconfigured: https://github.com/mackensen/moodle-hat
+- Docker containers for Moodle Developers and Behat: https://github.com/moodlehq/moodle-docker
+- Docker environment with Behat preconfigured : https://github.com/tobiga/docker_moodle_environment
+
+## Examples
+
+### Quick setup and testing using moodle-docker
+
+This is a quick guide to help locally pass tests for your developments, before submitting them:
+
+1. Set up a default Moodle install using [moodle-docker](https://github.com/moodlehq/moodle-docker), with the database and Moodle version of your choice. See its README for more details. This will start some docker containers.
+2. Initialize behat to start testing with this command, from the webserver container:
+
+   ```
+      php admin/tool/behat/cli/init.php
+   ```
+
+3. Run the behat test of your choice, from the webserver container. For instance:
+
+   ```
+      vendor/bin/behat --config /var/www/behatdata/behatrun/behat/behat.yml --tags tool_task
+   ```
+
+And you'll see something like:
+
+```php
+Moodle 4.0dev (Build: 20210507), 0b47ea0a44a092f9000729ca7b15fff23111538b
+Php: 7.3.26, mysqli: 5.7.21, OS: Linux 5.4.0-66-generic x86_64
+Run optional tests:
+
+Accessibility: No
+Server OS "Linux", Browser: "firefox"
+Started at 09-05-2021, 06:00
+...................................................................... 70
+...............................
+12 scenarios (12 passed)
+101 steps (101 passed)
+2m53.27s (47.61Mb)
+```
+
+## See also
+
+- [Acceptance testing for the mobile app](https://docs.moodle.org/dev/Acceptance_testing_for_the_mobile_app)
+
+## Information to re-home
```

Comment:
I would suggest creating a separate page for this section... Otherwise, we'll probably miss/forget to move it to the right section ;-)

Reply:
Apart from that, the "**Note:**" should be moved to admonitions ;-)